### ISSUE TITLE
Quantizers are not DDP/AMP compliant

### ISSUE THREAD

**Issue by danieltudosiu** [lucidrains/vector-quantize-pytorch#10]
Hi Lucidrains,

Thanks for the amazing work you do by implementing all those papers!

Is there a plan to make the Quantizer be compliant with:

- DDP - They need an all gather before calculating anything so the updates are exactly the same across all ranks 
- AMP - In my experience, if AMP touches upon the quantizers it screws up the gradient magnitudes making it NaN/Overflow

If you want I can have a go at it.

**Comment by lucidrains** [comment]
@danieltudosiu Hi Daniel! No that would be great! Always welcoming contributors :)

**Comment by lucidrains** [comment]
@danieltudosiu do you want to see if https://github.com/lucidrains/vector-quantize-pytorch/releases/tag/0.4.8 fixes the AMP issue?

**Comment by lucidrains** [comment]
as for DDP, i'm guessing just need an allreduce at these two lines? https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/vector_quantize_pytorch.py#L153-L155

**Comment by danieltudosiu** [comment]
> as for DDP, i'm guessing just need an allreduce at these two lines? https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/vector_quantize_pytorch.py#L153-L155

One reduction should happen [here](https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/vector_quantize_pytorch.py#L153) but only for the summation of the one hot encoddings (embed_onehot.sum(0)).

And one [here](https://github.com/lucidrains/vector-quantize-pytorch/blob/537fd0d73045477b74edebf449c45877f1458acb/vector_quantize_pytorch/vector_quantize_pytorch.py#L154) for the summation of the embeddings (embed_sum).

> @danieltudosiu do you want to see if https://github.com/lucidrains/vector-quantize-pytorch/releases/tag/0.4.8 fixes the AMP issue?

Regarding the AMP part, I am not actively using this codebase since we are close to finishing the project and we have a more barebone implementation ourselves, I was just signalling the issues so after the project I can move to this library ;) .

**Comment by lucidrains** [comment]
@danieltudosiu got it! thanks for your input :)

**Comment by danieltudosiu** [comment]
@lucidrains  just to be clear the all reduce should be something like this:

```
    import torch.distributed.distributed_c10d as dist

    if dist.is_initialized():
        dist.all_reduce(tensor=encodings_sum, op=dist.ReduceOp.SUM)
        dist.all_reduce(tensor=dw, op=dist.ReduceOp.SUM)
```

Where encodings_sum is your embed_onehot.sum(0) and dw is your embed_sum.

**Comment by lucidrains** [comment]
@danieltudosiu hey yup! i think sum is by default anyways :)

https://github.com/lucidrains/vector-quantize-pytorch#ddp
